presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -0,0 +1,481 @@
+                boolean splittable = isSplittable(files.getInputFormat(), hdfsEnvironment.getFileSystem(file.getPath()), file.getPath());
+
+                hiveSplitSource.addToQueue(createHiveSplits(
+                        files.getPartitionName(),
+                        file.getPath().toString(),
+                        file.getBlockLocations(),
+                        0,
+                        file.getLen(),
+                        files.getSchema(),
+                        files.getPartitionKeys(),
+                        splittable,
+                        session,
+                        files.getEffectivePredicate()));
+                if (hiveSplitSource.isQueueFull()) {
+                    fileIterators.addFirst(files);
+                    return;
+                }
+            }
+        }
+
+        // No need to put the iterator back, since its either empty or we've stopped
+    }
+
+    private void loadPartition(HivePartitionMetadata partition)
+            throws IOException
+    {
+        String partitionName = partition.getHivePartition().getPartitionId();
+        Properties schema = getPartitionSchema(table, partition.getPartition());
+        List<HivePartitionKey> partitionKeys = getPartitionKeys(table, partition.getPartition());
+        TupleDomain<HiveColumnHandle> effectivePredicate = partition.getHivePartition().getEffectivePredicate();
+
+        Path path = new Path(getPartitionLocation(table, partition.getPartition()));
+        Configuration configuration = hdfsEnvironment.getConfiguration(path);
+        InputFormat<?, ?> inputFormat = getInputFormat(configuration, schema, false);
+
+        if (inputFormat instanceof SymlinkTextInputFormat) {
+            JobConf jobConf = new JobConf(configuration);
+            FileInputFormat.setInputPaths(jobConf, path);
+            InputSplit[] splits = inputFormat.getSplits(jobConf, 0);
+
+            // TODO: This should use an iterator like the HiveFileIterator
+            for (InputSplit rawSplit : splits) {
+                FileSplit split = ((SymlinkTextInputFormat.SymlinkTextInputSplit) rawSplit).getTargetSplit();
+
+                // get the filesystem for the target path -- it may be a different hdfs instance
+                FileSystem targetFilesystem = hdfsEnvironment.getFileSystem(split.getPath());
+                FileStatus file = targetFilesystem.getFileStatus(split.getPath());
+                hiveSplitSource.addToQueue(createHiveSplits(
+                        partitionName,
+                        file.getPath().toString(),
+                        targetFilesystem.getFileBlockLocations(file, split.getStart(), split.getLength()),
+                        split.getStart(),
+                        split.getLength(),
+                        schema,
+                        partitionKeys,
+                        false,
+                        session,
+                        effectivePredicate));
+                if (stopped) {
```

Comment:
I think this is wrong. This loop is never resumed, so we would miss the rest of the files.